# Ch04_Q13(e, f, g, h, i, j)

## This question should be answered using the Weekly data set, which is part of the ISLP package. This data is similar in nature to the Smarket data from this chapter’s lab, except that it contains 1089 weekly returns for 21 years, from the beginning of 1990 to the end of 2010.

In [1]:
!pip install ISLP

  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 60.4/60.4 kB 2.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.6/3.6 MB 41.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 349.3/349.3 kB 18.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 522.0/522.0 kB 24.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 36.4/36.4 MB 22.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 815.2/815.2 kB 28.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 890.6/890.6 kB 22.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 94.5/94.5 kB 3.2 MB/s eta 0:00:00
  Created wheel for autograd-gamma: filename=autograd_gamma-0.5.0-py3-none-any.whl size=4031 sha256=bea5280799c1bcd9c8f0fe4a1cb21823483eacf4e420783da657b4896e6312ed
  Stored in directory: /root/.cache/pip/wheels/25/cc/e0/ef2969164144c899fedb22b338f6703e2b9cf46eeebf254991
Successfully built autograd-gamma
  Attempting un

In [2]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from ISLP import load_data
from sklearn.metrics import confusion_matrix, accuracy_score
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression

weekly = load_data('Weekly')
train_data = weekly[(weekly['Year'] >= 1990) & (weekly['Year'] <= 2008)]
test_data = weekly[(weekly['Year'] >= 2009) & (weekly['Year'] <= 2010)]
X_train = train_data[['Lag2']]
y_train = train_data['Direction']
X_test = test_data[['Lag2']]
y_test = test_data['Direction']

## (e) Repeat (d) using LDA

In [3]:
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis

# Train LDA model
lda_model = LinearDiscriminantAnalysis()
lda_model.fit(X_train, y_train)

# Make predictions on test data
lda_predictions = lda_model.predict(X_test)

# Calculate confusion matrix and accuracy
lda_conf_matrix = confusion_matrix(y_test, lda_predictions)
print("Confusion Matrix (LDA):\n", lda_conf_matrix)

lda_accuracy = accuracy_score(y_test, lda_predictions)
print("\nOverall Fraction of Correct Predictions (LDA):", lda_accuracy)


Confusion Matrix (LDA):
 [[ 9 34]
 [ 5 56]]

Overall Fraction of Correct Predictions (LDA): 0.625


## (f) Repeat (d) using QDA.

In [4]:
from sklearn.discriminant_analysis import QuadraticDiscriminantAnalysis

# Train QDA model
qda_model = QuadraticDiscriminantAnalysis()
qda_model.fit(X_train, y_train)

# Make predictions on test data
qda_predictions = qda_model.predict(X_test)

# Calculate confusion matrix and accuracy
qda_conf_matrix = confusion_matrix(y_test, qda_predictions)
print("Confusion Matrix (QDA):\n", qda_conf_matrix)

qda_accuracy = accuracy_score(y_test, qda_predictions)
print("\nOverall Fraction of Correct Predictions (QDA):", qda_accuracy)


Confusion Matrix (QDA):
 [[ 0 43]
 [ 0 61]]

Overall Fraction of Correct Predictions (QDA): 0.5865384615384616


## (g) Repeat (d) using KNN with K = 1.

In [6]:
from sklearn.neighbors import KNeighborsClassifier

# Train KNN model with K = 1
knn_model = KNeighborsClassifier(n_neighbors=1)
knn_model.fit(X_train, y_train)

# Make predictions on test data
knn_predictions = knn_model.predict(X_test)

# Calculate confusion matrix and accuracy
knn_conf_matrix = confusion_matrix(y_test, knn_predictions)
print("Confusion Matrix (KNN, K=1):\n", knn_conf_matrix)

knn_accuracy = accuracy_score(y_test, knn_predictions)
print("\nOverall Fraction of Correct Predictions (KNN, K=1):", knn_accuracy)


Confusion Matrix (KNN, K=1):
 [[22 21]
 [32 29]]

Overall Fraction of Correct Predictions (KNN, K=1): 0.49038461538461536


## (h)  Repeat (d) using naive Bayes.

In [7]:
from sklearn.naive_bayes import GaussianNB

# Train Naive Bayes model
nb_model = GaussianNB()
nb_model.fit(X_train, y_train)

# Make predictions on test data
nb_predictions = nb_model.predict(X_test)

# Calculate confusion matrix and accuracy
nb_conf_matrix = confusion_matrix(y_test, nb_predictions)
print("Confusion Matrix (Naive Bayes):\n", nb_conf_matrix)

nb_accuracy = accuracy_score(y_test, nb_predictions)
print("\nOverall Fraction of Correct Predictions (Naive Bayes):", nb_accuracy)

Confusion Matrix (Naive Bayes):
 [[ 0 43]
 [ 0 61]]

Overall Fraction of Correct Predictions (Naive Bayes): 0.5865384615384616


## (i)  Which of these methods appears to provide the best results on this data?

### The Linear Discriminant Analysis (LDA) model provides the highest accuracy on this dataset, with an accuracy of 62.5%. Therefore, LDA appears to provide the best results on this data.

## (j)  Experiment with diferent combinations of predictors, including possible transformations and interactions, for each of the methods. Report the variables, method, and associated confusion matrix that appears to provide the best results on the held out data. Note that you should also experiment with values for K in the KNN classifer.

In [11]:
X_train_full = train_data[['Lag1', 'Lag2', 'Lag3', 'Lag4', 'Lag5']]
X_test_full = test_data[['Lag1', 'Lag2', 'Lag3', 'Lag4', 'Lag5']]

# Loop over different values of K
for k in [1, 3, 5, 7]:
    knn_model = KNeighborsClassifier(n_neighbors=k)
    knn_model.fit(X_train_full, y_train)
    knn_predictions = knn_model.predict(X_test_full)
    print(f"Accuracy (KNN, K={k}):", accuracy_score(y_test, knn_predictions))


Accuracy (KNN, K=1): 0.5192307692307693
Accuracy (KNN, K=3): 0.5576923076923077
Accuracy (KNN, K=5): 0.5480769230769231
Accuracy (KNN, K=7): 0.5673076923076923
